File: puzzle_15.c
#include <stdio.h>
#include <stdlib.h>
#include <stdio_ext.h>
#include <math.h>
#include <string.h>

// ========================== Estruturas ========================== 
typedef struct {
  int i, j;
}posicoes;

// ========================== Variaveis globais ========================== 
int in[4][4];
int tabuleiro[16];
posicoes pecas_corretas[16];

// ========================== Funcoes gerais ========================== 
void organiza_posicoes_pecas(){
  pecas_corretas[0].i = 2; 
  pecas_corretas[0].j = 1; 
  
  pecas_corretas[1].i = 0; 
  pecas_corretas[1].j = 0; 
  
  pecas_corretas[2].i = 0; 
  pecas_corretas[2].j = 1; 
  
  pecas_corretas[3].i = 0; 
  pecas_corretas[3].j = 2; 
  
  pecas_corretas[4].i = 0; 
  pecas_corretas[4].j = 3; 
  
  pecas_corretas[5].i = 1; 
  pecas_corretas[5].j = 3; 
  
  pecas_corretas[6].i = 2; 
  pecas_corretas[6].j = 3; 
  
  pecas_corretas[7].i = 3; 
  pecas_corretas[7].j = 3; 
  
  pecas_corretas[8].i = 3; 
  pecas_corretas[8].j = 2; 
  
  pecas_corretas[9].i = 3; 
  pecas_corretas[9].j = 1; 
  
  pecas_corretas[10].i = 3; 
  pecas_corretas[10].j = 0; 
  
  pecas_corretas[11].i = 2; 
  pecas_corretas[11].j = 0; 
  
  pecas_corretas[12].i = 1; 
  pecas_corretas[12].j = 0; 
  
  pecas_corretas[13].i = 1; 
  pecas_corretas[13].j = 1; 
  
  pecas_corretas[14].i = 1; 
  pecas_corretas[14].j = 2; 
  
  pecas_corretas[15].i = 2; 
  pecas_corretas[15].j = 2; 
  
}

int max_3(int v1, int v2, int v3){
  int maior;
  
  if (v1 > v2 && v1 > v3){
    maior = v1;
  } else if (v2 > v3){
    maior = v2;
  } else {
    maior = v3;
  }

  return maior;
}

int distancia_manhatan(int valor, int i, int j){
  int i_correto, j_correto, soma;

  soma = 0;

  i_correto = pecas_corretas[valor].i;
  j_correto = pecas_corretas[valor].j;

  if (i > i_correto){
    soma += i - i_correto;
  } else {
    soma += i_correto - i;
  }

  if (j > j_correto){
    soma += j - j_correto;
  } else {
    soma += j_correto - j;
  }

  return soma;
}

void organiza_espiral(int entrada[][4]){
  int j;

  j = 0;

  for(j = 0; j<4; j++){
    tabuleiro[j] = entrada[0][j];
    tabuleiro[3+j] = entrada[j][3];
    tabuleiro[9-j] = entrada[3][j];
  }

  for(j=0;j<3;j++){
    tabuleiro[11+j] = entrada[1][j];
    tabuleiro[11-j] = entrada[j+1][0];
  }

  tabuleiro[14] = entrada[2][2];
  tabuleiro[15] = entrada[2][1];

}

void le_tabuleiro(){

  scanf("%d %d %d %d %d %d %d %d %d %d %d %d %d %d %d %d",
            &in[0][0], &in[0][1], &in[0][2], &in[0][3],
            &in[1][0], &in[1][1], &in[1][2], &in[1][3],
            &in[2][0], &in[2][1], &in[2][2], &in[2][3],
            &in[3][0], &in[3][1], &in[3][2], &in[3][3]);

  organiza_espiral(in);
}

// ========================== Heuristicas ========================== 
int h_linha_1(){
  int tabuleiro_correto[16] = {1,2,3,4,5,6,7,8,9,10,11,12,13,14,15, 0};
  int pecas_fora, i;

  pecas_fora = i = 0;

  while(i < 16){
    if(tabuleiro[i] != tabuleiro_correto[i]){
      if (tabuleiro[i] != 0){
        pecas_fora++;
      }
    }
    i++;
  }

  return pecas_fora;
}

int h_linha_2(){
  int pecas_fora, i;

  pecas_fora = 0;
  i = 1;
  while(i <16){
    if (tabuleiro[i-1] == 0){ i++;}
    if(tabuleiro[i] != (tabuleiro[i-1]+1)){
      pecas_fora++;
    }
    i++;
  }

  return pecas_fora;
}

int h_linha_3(){
  int i, j, soma;

  soma = 0;

  for(i=0; i<4; i++){
    for(j=0; j<4; j++){
      if (in[i][j] != 0){
        soma += distancia_manhatan(in[i][j], i, j);
      }
    }
  }

  return soma;
}

int h_linha_4(){
  return 0;
}

int h_linha_5(){
  return max_3(h_linha_1(), h_linha_2(), h_linha_3());
}

// ========================== In/Out ========================== 
void print_inline_tabuleiro(){
  int i;

  printf("\nValores em espiral:\n");
  for(i=0; i<16;i++){
    printf("%d ", tabuleiro[i]);
  }
  printf("\n");
}

void print_entrada(int entrada[][4]){
  int i,j;

  i = j = 0;

  printf("\nEntrada lida:\n");

  for(i=0; i<4;i++){
    for (j=0; j<4; j++) {
      printf("%d ", entrada[i][j]);
    }
    printf("\n");
  }
}

//  ========================== main  ========================== 
int main(){
  //int h1;
  //int h2;
  int h3;

  organiza_posicoes_pecas();
  le_tabuleiro();
  //print_inline_tabuleiro();

  //h1 = h_linha_1();
  //printf("%d", h1);

  //h2 = h_linha_2();
  //printf("%d", h2);

  h3 = h_linha_3();
  printf("%d", h3);

  return 0;
}

File: puzzle_15.py
import math, copy, heapq

################################ Classe para guardar o tabuleiro################################
class entradaAStar: 
    def __init__(self, matriz, g, h, pai, sucessores):
        self.matriz = copy.deepcopy(matriz)
        self.g = g
        self.h = h
        self.pai = pai
        self.sucessores = sucessores

    def f(self):
        return self.g + self.h
    
    def add_sucessor(self, sucessor):
        self.sucessores.append(sucessor)

######################################## Variaveis Globais ########################################
pecas_corretas = [
        [2,1], [0,0], [0,1], [0,2],
        [0,3], [1,3], [2,3], [3,3],
        [3,2], [3,1], [3,0], [2,0],
        [1,0], [1,1], [1,2], [2,2]]

######################################## Funcoes Gerais ########################################
def distancia_manhatan(valor, i, j):
    soma = 0

    i_correto = pecas_corretas[valor][0]
    j_correto = pecas_corretas[valor][1]

    soma += int(math.fabs(abs(j - j_correto)))
    soma += int(math.fabs(abs(i - i_correto)))

    return soma

def organiza_espiral(entrada, tabuleiro):
    j = 0

    for j in range(0,4):
        tabuleiro[j] = entrada[0][j]
        tabuleiro[3+j] = entrada[j][3]
        tabuleiro[9-j] = entrada[3][j]

    for j in range(0,3):
        tabuleiro[11+j] = entrada[1][j]
        tabuleiro[11-j] = entrada[j+1][0]

    tabuleiro[14] = entrada[2][2]
    tabuleiro[15] = entrada[2][1]

def le_tabuleiro(entrada, tabuleiro):
    ent1 = input().split(" ")
    ent1 = [x for x in ent1 if x]
    z = 0

    for i in range(0,4):
        for j in range(0,4):
            entrada[i][j] = int(ent1[z])
            z+=1

    organiza_espiral(entrada, tabuleiro)

def acha_sucessores(node):
    matriz = copy.deepcopy(node.matriz)
    condicao = False
    i = j = 0

    while (not condicao):
        if j == 4:
            j = 0
            i += 1
        if matriz[i][j] == 0:
            condicao = True
        else:
            j +=1

    if ((i-1)>=0):
        matriz1 = copy.deepcopy(node.matriz)
        # troca a peca de lugar com o 0
        matriz1[i][j], matriz1[i-1][j] = matriz1[i-1][j], matriz1[i][j]
        # adiciona nos filhos do node
        node.add_sucessor(entradaAStar(matriz1,node.g +1,0, [], []))

    if ((i+1)<=3):
        matriz2 = copy.deepcopy(node.matriz)
        matriz2[i][j], matriz2[i+1][j] = matriz2[i+1][j], matriz2[i][j]
        node.add_sucessor(entradaAStar(matriz2,node.g +1,0, [], []))

    if ((j-1)>=0):
        matriz3 = copy.deepcopy(node.matriz)
        matriz3[i][j], matriz3[i][j-1] = matriz3[i][j-1], matriz3[i][j]
        node.add_sucessor(entradaAStar(matriz3,node.g +1,0, [],[]))

    if ((j+1)<=3):
        matriz4 = copy.deepcopy(node.matriz)
        matriz4[i][j], matriz4[i][j+1] = matriz4[i][j+1], matriz4[i][j]
        node.add_sucessor(entradaAStar(matriz4,node.g +1,0, [], []))

def findMenor(A):
    valores = list(A.values())
    v = valores[0].f()
    menor = valores[0]

    for elem in valores:
        if (elem.f() < v):
            v = elem.f()
            menor = elem

    return str(menor.matriz)

######################################## Heuristicas ########################################
def h_linha_1(entrada):
    tabuleiro_correto = [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,0]
    tabuleiro = list(range(16))

    pecas_fora = i = 0
    organiza_espiral(entrada, tabuleiro)

    while(i < 16):
        if(tabuleiro[i] != tabuleiro_correto[i]):
            if (tabuleiro[i] != 0):
                pecas_fora+=1
        i+=1

    return pecas_fora

def h_linha_2(entrada):
    tabuleiro = list(range(16))
    pecas_fora = 0
    i = 1

    organiza_espiral(entrada, tabuleiro)

    while(i <15):
        if (tabuleiro[i-1] == 0):
            i+=1
        if(tabuleiro[i] != (tabuleiro[i-1]+1)):
            pecas_fora+=1
        i+=1

    return pecas_fora

def h_linha_3(entrada):
    soma = 0

    for i in range(0, 4):
        for j in range(0, 4):
            if (entrada[i][j] != 0):
                soma += distancia_manhatan(entrada[i][j], i, j)

    return soma

def h_linha_4(entrada):
    p = [0.4, 0.2, 0.4];

    return int(p[0]*h_linha_1(entrada) + p[1]*h_linha_2(entrada) + p[2]*h_linha_3(entrada))


def h_linha_5(entrada):
    return max(h_linha_1(entrada), h_linha_2(entrada), h_linha_3(entrada))


######################################### A* #########################################
def AStar(start):
    matriz_final = [[1,2,3,4],[12,13,14,5],[11,0,15,6],[10,9,8,7]]
    T = str(matriz_final)
    A = {}          # Dicionario: {"key" : "objeto"}
                    # https://docs.python.org/3/library/stdtypes.html#dict
    
    # A <- S
    A[str(start)] = entradaAStar(start,0,h_linha_5(start), [], [])
    # F <- 0
    F = {}

    h = []
    heapq.heappush(h, (A.get(str(start)).f(), str(start)))

    v = A.get(str(start))
    #while A:
    while A and (str(v.matriz) != T):
        # v existe em A, tal que, f(v) = min {f(v)}
        #v = A.get(findMenor(A))
        condicao = True
        while condicao:
            posicao = heapq.heappop(h)
            if posicao[1] in A:
                condicao = False
        v = A.get(posicao[1])
        #print("V: " + str(v.matriz) + " g: "+ str(v.g))

        # A <- A - {v}
        A.pop(str(v.matriz))
        # F <- F U {v}
        F[str(v.matriz)] = v

        # Para cada m(e)Gamma(v)
        if not v.sucessores:
            acha_sucessores(v)
        m = v.sucessores
        for i in range(0, len(v.sucessores)):
            # calcule g(m)
            #m[i].g += 1 (ja feito ao achar o sucessor)
            #print("Filho #" + str(i+1)+": " + str(m[i].matriz) + " g: "+str(m[i].g))

            # Se existe m' em A, tal que 
            # m'=m e g(m)<g(m')
            m_linha = str(m[i].matriz)
            if (m_linha in A) and (m[i].g < A[m_linha].g):
                # A <- A - {m'}
                A.pop(m_linha)

            # se m nao existe em A(U)F
            if (m_linha not in A) and (m_linha not in F):
                # A <- A(U){m}
                A[str(m[i].matriz)] = m[i]
                A[str(m[i].matriz)].pai = v;
                A[str(m[i].matriz)].h = h_linha_5(m[i].matriz)

                #heap para achar o menor
                heapq.heappush(h, (A[str(m[i].matriz)].f(), str(m[i].matriz)))

    return (v.g)

######################################## Main ########################################
def main():
    entrada = [[0,0,0,0],[0,0,0,0],[0,0,0,0],[0,0,0,0]]
    tabuleiro = list(range(16))

    le_tabuleiro(entrada, tabuleiro)
  
    #print(h_linha_1(entrada))
    #print(h_linha_2(entrada))
    #print(h_linha_3(entrada))
    #print(h_linha_4(entrada))
    #print(h_linha_5(entrada))
    print(AStar(entrada))


if __name__ == '__main__':
    main()
